### PR TITLE
`ProxyConnectConnectionFactoryFilter`: report when the channel is closed

### PR DESCRIPTION
**Pull request by idelpivnitskiy**
Motivation:

When proxy returns 2xx status code with `Connection: closed` header, users see an awkward error saying:
Failed to find a handler of type DeferSslHandler in channel pipeline. It confuses users. We can handle this case with a better exception message.

Modifications:
- Check if the channel is still active before generating `IllegalStateException`;

Result:

More clear exception message when proxy closes the connection.

### CODE REVIEW

File: servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/ProxyConnectConnectionFactoryFilter.java [hunk 1]
```diff
@@ -86,7 +87,7 @@ public Single<C> newConnection(final ResolvedAddress resolvedAddress,
                 return delegate().newConnection(resolvedAddress, contextMap, observer).flatMap(c -> {
                     try {
                         return c.request(c.connect(connectAddress).addHeader(CONTENT_LENGTH, ZERO))
-                                .flatMap(response -> handleConnectResponse(c, response))
+                                .flatMap(response -> handleConnectResponse(c, response).shareContextOnSubscribe())
```
Comment (daschl) [inline servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/ProxyConnectConnectionFactoryFilter.java:90]: for my own education: is this related to the changes below and if so why?

File: servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/ProxyConnectConnectionFactoryFilter.java [hunk 2]
```diff
@@ -123,7 +124,14 @@ public void userEventTriggered(final ChannelHandlerContext ctx, final Object evt
 
             final DeferSslHandler deferSslHandler = channel.pipeline().get(DeferSslHandler.class);
             if (deferSslHandler == null) {
-                return failed(new IllegalStateException("Failed to find a handler of type " +
+                if (!channel.isActive()) {
+                    LOGGER.info("{} is unexpectedly closed after receiving response: {}",
```
Comment (daschl) [inline servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/ProxyConnectConnectionFactoryFilter.java:128]: Is there a hint we can give to the user that this is proxy-related?

**Review by daschl — APPROVED** [review]
Looks good to me, just two comments.

File: servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/ProxyConnectConnectionFactoryFilter.java (hunk 1 above)
Comment (idelpivnitskiy) [reply servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/ProxyConnectConnectionFactoryFilter.java:90]: no, just observed as part of the main fix. The `flatMap` is like `defer`, will make a copy of a context when it subscribes to a returned single. Because `handleConnectResponse` is 100% internal method (users can not do anything in the middle of that method) it doesn't matter if we copy the context or share. Added `shareContextOnSubscribe()` to avoid unnecessary copying, just to reduce allocations.
